modules/authwindowslive/lib/Auth/Source/LiveID.php

```diff
@@ -96,38 +100,32 @@ public function finalStep(&$state) {
 			),
 		);
 
-		$result = \SimpleSAML\Utils\HTTP::fetch('https://consent.live.com/AccessToken.aspx', $context);
+		$result = \SimpleSAML\Utils\HTTP::fetch('https://login.microsoftonline.com/common/oauth2/v2.0/token', $context);
 
-		parse_str($result, $response);
+		$response = json_decode($result,true);
 
 		// error checking of $response to make sure we can proceed
-		if (!array_key_exists('wrap_access_token',$response))
-			throw new Exception('[' . $response['error_code'] . '] ' . $response['wrap_error_reason'] .
-				"\r\nNo wrap_access_token returned - cannot proceed\r\n" . $response['internal_info']);
+		if (!array_key_exists('access_token',$response))
+			throw new Exception('[' . $response['error'] . '] ' . $response['error_description'] .
+				"\r\nNo access_token returned - cannot proceed\r\n" . implode(', ', $response['error_codes']));
 
-		SimpleSAML\Logger::debug("Got an access token from the OAuth WRAP service provider [" .
-			$response['wrap_access_token'] . "] for user [" . $response['uid'] . "]");
+		SimpleSAML\Logger::debug("authwindowslive: Got an access token from the OAuth service provider [" .
+			$response['access_token'] . "]");
 
-		// Documentation at: http://msdn.microsoft.com/en-us/library/ff751708.aspx
-		$opts = array('http' => array('header' => "Accept: application/json\r\nAuthorization: WRAP access_token=" .
-						$response['wrap_access_token'] . "\r\n"));
-		$data = \SimpleSAML\Utils\HTTP::fetch('https://apis.live.net/V4.1/cid-'. $response['uid'] . '/Profiles',$opts);
+		// Documentation at: http://graph.microsoft.io/en-us/docs/overview/call_api
+		$opts = array('http' => array('header' => "Accept: application/json\r\nAuthorization: Bearer " .
+						$response['access_token'] . "\r\n"));
+		$data = \SimpleSAML\Utils\HTTP::fetch('https://graph.microsoft.com/v1.0/me',$opts);
 		$userdata = json_decode($data, TRUE);
 
+		// This is the simplest case
+		if(!array_key_exists('@odata.context',$userdata) OR array_key_exists('error',$userdata))
+			throw new Exception('Unable to retrieve userdata from Microsoft Graph [' . $userdata['error']['code'] . '] ' . $userdata['error']['message']); 
 		$attributes = array();
-		$attributes['windowslive_uid'] = array($response['uid']);
-		$attributes['windowslive_targetedID'] = array('http://windowslive.com!' . $response['uid']);
-		$attributes['windowslive_user'] = array($response['uid'] . '@windowslive.com');
-
-		if (array_key_exists('Entries',$userdata)) {
-			foreach($userdata['Entries'][0] AS $key => $value) {
-				if (is_string($value))
-					$attributes['windowslive.' . $key] = array((string)$value);
-			}
-
-			if (array_key_exists('Emails', $userdata['Entries'][0]))
-				$attributes['windowslive_mail'] = array($userdata['Entries'][0]['Emails'][0]['Address']);
-
+		$attributes['windowslive_targetedID'] = array('https://graph.microsoft.com!' . (!empty($userdata['id']) ? $userdata['id'] : 'unknown'));
+		foreach($userdata AS $key => $value) {
```

Comment:
Thanks for the feedback - I'm offline for the next week or so, but I'll
look at all of this when I get back to somewhere with reliable connectivity.

On 3 June 2016 at 09:39, Jaime Pérez Crespo <notifications@github.com>
wrote:

> In modules/authwindowslive/lib/Auth/Source/LiveID.php
> <https://github.com/simplesamlphp/simplesamlphp/pull/398#discussion_r65667356>
> :
>
> >  		$attributes = array();
> > -		$attributes['windowslive_uid'] = array($response['uid']);
> > -		$attributes['windowslive_targetedID'] = array('http://windowslive.com!' . $response['uid']);
> > -		$attributes['windowslive_user'] = array($response['uid'] . '@windowslive.com');
> > -
> > -		if (array_key_exists('Entries',$userdata)) {
> > -			foreach($userdata['Entries'][0] AS $key => $value) {
> > -				if (is_string($value))
> > -					$attributes['windowslive.' . $key] = array((string)$value);
> > -			}
> > -
> > -			if (array_key_exists('Emails', $userdata['Entries'][0]))
> > -				$attributes['windowslive_mail'] = array($userdata['Entries'][0]['Emails'][0]['Address']);
> > -
> > +		$attributes['windowslive_targetedID'] = array('https://graph.microsoft.com!' . (!empty($userdata['id']) ? $userdata['id'] : 'unknown'));
> > +		foreach($userdata AS $key => $value) {
>
> Keywords should always be in lowercase, according to the PSR-2 coding
> standard that we are using
> <https://github.com/simplesamlphp/simplesamlphp/blob/master/CONTRIBUTE.md>
> .
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/simplesamlphp/simplesamlphp/pull/398/files/7cb89f9a3f229c7a21d94677c264bb70fe29cbf4#r65667356>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AHoE2aVQhqaRLvMesmjEqpGCu6Znbrkwks5qH9okgaJpZM4IrOlu>
> .
>